_internal/utils/preload.ts

```diff
@@ -0,0 +1,32 @@
+import { Middleware, Key, BareFetcher } from '../types'
+import { serialize } from './serialize'
+
+const preloadMap = new Map<string, Promise<any>>()
+
+export const preload = (key: Key, fetcher: BareFetcher<any>) => {
+  const promise = fetcher(key)
+  const keyString = serialize(key)[0]
+  preloadMap.set(keyString, promise)
+  return promise
+}
+
+export const middleware: Middleware = useSWRNext => (key, fetcher_, config) => {
+  const fetcher =
+    fetcher_ == null
+      ? null
+      : // fetcher might be a sync function, so this should not be an async function
```

Comment:
If this is an async function, https://github.com/vercel/swr/blob/9263ca79f0747f9132fdcd0cf5eb0e47f7ee08a8/test/use-swr-error.test.tsx#L433 is failed.

Reply:
Makes sense!